torch/csrc/jit/passes/peephole_list_idioms.cpp

```diff
@@ -0,0 +1,104 @@
+#include <ATen/core/jit_type.h>
+#include <torch/csrc/jit/ir/alias_analysis.h>
+#include <torch/csrc/jit/ir/ir_views.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/passes/peephole.h>
+#include <torch/csrc/jit/runtime/graph_executor.h>
+#include <torch/csrc/utils/memory.h>
+
+namespace torch {
+namespace jit {
+
+size_t normalizeIndex(int64_t index, size_t len) {
+  if (index < 0) {
+    return index + len;
+  } else {
+    return index;
+  }
+}
+
+// This pass only does optimizations on lists which aren't mutated,
+// so we first use the Alias Db to collect the set of list values
+// which we shouldn't optimize.
+struct PeepholeOptimizeListIdiomsImpl {
+  PeepholeOptimizeListIdiomsImpl(const std::shared_ptr<Graph>& graph)
+      : graph_(graph), aliasDb_(torch::make_unique<AliasDb>(graph_)) {
+    collectMutatedLists(graph_->block());
+    run(graph_->block());
+  }
+
+ private:
+  void checkForMutatedList(Value* v) {
+    if (v->type()->cast<ListType>() && aliasDb_->hasWriters(v)) {
+      mutated_lists_.insert(v);
+    }
+  }
+
+  void collectMutatedLists(Block* b) {
+    for (Value* v : b->inputs()) {
+      checkForMutatedList(v);
+    }
+    for (Node* n : b->nodes()) {
+      for (Value* v : n->outputs()) {
+        checkForMutatedList(v);
+      }
+      for (Block* block : n->blocks()) {
+        collectMutatedLists(block);
+      }
+    }
+  }
+
+  void run(Block* block) {
+    for (Node* node : block->nodes()) {
+      for (Block* b : node->blocks()) {
+        run(b);
+      }
+
+      // only optimizing list ops
+      if (node->inputs().size() == 0 ||
```

Comment:
Seems like it would be clearer to filter here based on node kind. We only optimize `aten::len` and `aten::__getitem__` so it seems like this more general check is unnecessary.

Reply:
aten::len and aten::__getitem__ are overloaded